devdoc/channel.md

```diff
@@ -0,0 +1,274 @@
+
+  - **SRS_CHANNEL_43_069: [** If the `state` has changed, `channel_pull` shall retry from requirement 43_010. **]**
+
+  - **SRS_CHANNEL_43_017: [** call `async_op_create` with `cancel_pull` as the cancel function. **]**
+
+  - **SRS_CHANNEL_43_018: [** store the given `pull_callback` and the given `pull_context` in the `channel`. **]**
+
+  - **SRS_CHANNEL_43_019: [** store `EPOCH` in the created `THANDLE(ASYNC_OP)`. **]**
+
+  - **SRS_CHANNEL_43_020: [** store the given `channel` in the created `THANDLE(ASYNC_OP)` by calling `THANDLE_INITIALIZE`. **]**
+
+  - **SRS_CHANNEL_43_021: [** store the created `THANDLE(ASYNC_OP)` in `out_op_pull` by calling `THANDLE_INITIALIZE_MOVE`. **]**
+
+  - **SRS_CHANNEL_43_062: [** set the `state` of the given `channel` to `CHANNEL_PULLED` by calling `interlocked_exchange`. **]**
+
+  - **SRS_CHANNEL_43_063: [** signal waiting threads by calling `wake_by_address_all`. **]**
+
+  - **SRS_CHANNEL_43_022: [** return `CHANNEL_RESULT_WILL_COMPLETE_ASYNC`. **]**
+
+**SRS_CHANNEL_43_023: [** If there are any failures, `channel_pull` shall fail and return `CHANNEL_RESULT_ERROR`. **]**
+
+### cancel_pull
+```
+static void cancel_pull(void* context)
+```
+
+`cancel_pull` is the cancel function given to `async_op_create` when `channel_pull` is called.
+
+**SRS_CHANNEL_43_046: [** atomically compare the current `EPOCH` with the `EPOCH` stored during `channel_pull`, increment the `EPOCH` and set the `state` of the `channel` stored in the given `context` to `CHANNEL_READY` if the `EPOCH` has not changed by calling `interlocked_compare_exchange`**]**
+
+**SRS_CHANNEL_43_048: [** If the `EPOCH` has not changed, `cancel_pull` shall call the `pull_callback` given to `channel_pull` with `pull_context` as the `pull_context` given to `channel_pull`, `data` as `NULL` and `result` as `CHANNEL_RESULT_CANCELLED`. **]**
+
+
+### channel_push
+```c
+    MOCKABLE_FUNCTION(, CHANNEL_RESULT, channel_push, THANDLE(CHANNEL), channel, THANDLE(RC_PTR), data, PUSH_CALLBACK, push_callback, void*, push_context, THANDLE(ASYNC_OP)*, out_op_push);
+```
+
+`channel_push` notifies the channel that there is data available and registers the given `push_callback` to be called when the given `data` has been consumed.
+
+**SRS_CHANNEL_43_024: [** If `channel` is `NULL`, `channel_push` shall fail and return `CHANNEL_RESULT_INVALID_ARGS`. **]**
+
+**SRS_CHANNEL_43_025: [** If `push_callback` is `NULL`, `channel_push` shall fail and return `CHANNEL_RESULT_INVALID_ARGS`. **]**
+
+**SRS_CHANNEL_43_026: [** If `out_op_push` is `NULL`, `channel_push` shall fail and return `CHANNEL_RESULT_INVALID_ARGS`. **]**
+
+**SRS_CHANNEL_43_027: [** `channel_push` shall call `interlocked_add` to obtain the current `state` of the `channel`. **]**
+
+**SRS_CHANNEL_43_028: [** If the obtained `state` is `CHANNEL_PUSHING` or `CHANNEL_PUSHED`, `channel_push` shall fail and return `CHANNEL_RESULT_REFUSED`. **]**
```

Comment:
>If the obtained `state` is `CHANNEL_PUSHING` or `CHANNEL_PUSHED`, [](http://example.com/codeflow?start=26&length=65)

So I am going to have a comment here that nobody will like (self included), but.

On one hand we are doing all the fancy interlocked thingies and smart bitness and everything. However, the result is this:

If the producer has 100 items, it can still only hand over these items ONE by ONE, back to back. That is, the next item, even when available, cannot be passed to `push` until the previous item has been consumed. The problem with this being of 2 types:

1) performance - back to back is not very quick... sigh

2) ease of use - as a producer looks like I would have to manage the state of this component myself - and make sure that I do not call "push" unless the state is somehow detected as "ready to receive a push".

====

I can hear arguments that say: "but we're going to have 222 of these in a <<queue>>" and I still believe that everything above still applies.

Maybe we should reconsider this design. (that's how back we'd have to go).

./waiting